src/Routing/Url.php

```diff
@@ -0,0 +1,271 @@
+<?php
+/**
+ * CakePHP(tm) : Rapid Development Framework (https://cakephp.org)
+ * Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ *
+ * Licensed under The MIT License
+ * For full copyright and license information, please see the LICENSE.txt
+ * Redistributions of files must retain the above copyright notice.
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ * @link          https://cakephp.org CakePHP(tm) Project
+ * @since         3.5.0
+ * @license       https://opensource.org/licenses/mit-license.php MIT License
+ */
+namespace Cake\Routing;
+
+use ArrayAccess;
+
+/**
+ * URL Object
+ *
+ * Use this object whenever you need to generate an URL for your app.
+ *
+ * This is not thought for external URLs.
+ */
+class Url implements ArrayAccess
+{
+
+    /**
+     * Array representation of the URL
+     *
+     * @var array
+     */
+    protected $url = [];
+
+    /**
+     * Absolute URL
+     *
+     * @var bool
+     */
+    protected $absolute;
+
+    /**
+     * Sets the action
+     *
+     * @param string $action Controller action name
+     * @return $this
+     */
+    public function setAction($action)
+    {
+        $this->url['action'] = $action;
+
+        return $this;
+    }
+
+    /**
+     * Sets the controller
+     *
+     * @param string $controller Controller class name without the controller suffix
+     * @return $this
+     */
+    public function setController($controller)
+    {
+        $this->url['controller'] = $controller;
+
+        return $this;
+    }
+
+    /**
+     * Set prefix
+     *
+     * @param string $prefix Prefix
+     * @return $this
+     */
+    public function setPrefix($prefix)
+    {
+        $this->url['prefix'] = $prefix;
+
+        return $this;
+    }
+
+    /**
+     * Sets the plugin
+     *
+     * @param string|null $plugin Plugin name
+     * @return $this
+     */
+    public function setPlugin($plugin)
+    {
+        $this->url['plugin'] = $plugin;
+
+        return $this;
+    }
+
+    /**
+     * Sets values to be passed in the URL
+     *
+     * @param string $pass Args to be passed to a controller action
+     * @return $this
+     */
+    public function setPass($pass)
```

Comment:
Should this be named `addPass`? `setPass` suggests that the old value would be overwritten.